src/projections/adams.cpp

```diff
@@ -0,0 +1,217 @@
+* PROJ by Kristian Evers. Original code found in file src/proj_guyou.c, see
+* https://github.com/rouault/libproj4/blob/master/libproject-1.01/src/proj_guyou.c
+* for reference.
+*
+* Copyright (c) 2005, 2006, 2009   Gerald I. Evenden
+* Copyright (c) 2020 Kristian Evers
+*
+* Related material
+* ----------------
+*
+*  CONFORMAL PROJECTION OF THE SPHERE WITHIN A SQUARE, 1929, OSCAR S. ADAMS,
+*  U.S.COASTAND GEODETIC SURVEY, Special Publication No.153,
+*  ftp://ftp.library.noaa.gov/docs.lib/htdocs/rescue/cgs_specpubs/QB275U35no1531929.pdf
+*
+*  https://en.wikipedia.org/wiki/Guyou_hemisphere-in-a-square_projection
+*  https://en.wikipedia.org/wiki/Adams_hemisphere-in-a-square_projection
+*  https://en.wikipedia.org/wiki/Peirce_quincuncial_projection
+*/
+
+#define PJ_LIB__
+
+#include <math.h>
+#include <errno.h>
+
+#include "proj.h"
+#include "proj_internal.h"
+
+PROJ_HEAD(guyou, "Guyou") "\n\tMisc Sph No inv";
+PROJ_HEAD(peirce_q, "Pierce Quincuncial") "\n\tMisc Sph No inv";
+PROJ_HEAD(adams_hemi, "Adams Hemisphere in a Square") "\n\tMisc Sph No inv";
+PROJ_HEAD(adams_wsI, "Adams World in a Square I") "\n\tMisc Sph No inv";
+PROJ_HEAD(adams_wsII, "Adams World in a Square II") "\n\tMisc Sph No inv";
+
+namespace { // anonymous namespace
+
+enum projection_type {
+    GUYOU,
+    PEIRCE_Q,
+    ADAMS_HEMI,
+    ADAMS_WSI,
+    ADAMS_WSII,
+};
+
+struct pj_opaque {
+    projection_type mode;
+};
+
+} // anonymous namespace
+
+
+#define TOL 1e-9
+#define RSQRT2 0.7071067811865475244008443620
+
+#define ORDER 8
+
+
+static double ell_int_5(double phi) {
+/* Procedure to compute elliptic integral of the first kind
+ * where k^2=0.5.  Precision good to better than 1e-7
+ * The approximation is performed with an even Chebyshev
+ * series, thus the coefficients below are the even values
+ * and where series evaluation  must be multiplied by the argument. */
+
+    int i = ORDER;
+    double d1 = 0.0;
+    double d2 = 0.0;
+    const double C[] = { /* even coefficients */
+        2.19174570831038,
+        0.0914203033408211,
+        -0.00575574836830288,
+        -0.0012804644680613,
+        5.30394739921063e-05,
+        3.12960480765314e-05,
+        2.02692115653689e-07,
+        -8.58691003636495e-07};
+    double const *Cp = C + ORDER - 1;
+
+    double y = phi * M_2_PI;
+    y = 2. * y * y - 1.;
+    double y2 = 2. * y;
+    while (--i) {
+        double temp = d1;
+        d1 = y2 * d1 - d2 + *Cp--;
+        d2 = temp;
+    }
+
+    return phi * (y * d1 - d2 + 0.5 * *Cp);
+}
+
+static PJ_XY forward(PJ_LP lp, PJ *P) {
+    double a=0., b=0.;
+    bool sm=false, sn=false;
+    PJ_XY xy;
+    struct pj_opaque *Q = static_cast<struct pj_opaque*>(P->opaque);
+
+    switch (Q->mode) {
+    case GUYOU:
+        if ((fabs(lp.lam) - TOL) > M_PI_2) {
+            proj_errno_set(P, PJD_ERR_TOLERANCE_CONDITION);
+        }
+
+        if (fabs(fabs(lp.phi) - M_PI_2) < TOL) {
+            xy.x = 0;
```

Comment:
Case not tested according to https://coveralls.io/builds/30051762/source?filename=src/projections/adams.cpp